presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1586,7 +1593,172 @@ public static Double stSphericalLength(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) S
+        double distanceVia180 = start - MIN_LONGITUDE + MAX_LONGITUDE - end;
+        if (distanceVia0 < distanceVia180) {
+            return liesInRange(longitude, start, end);
+        }
+        else {
+            return liesInRange(longitude, MIN_LONGITUDE, start) || liesInRange(longitude, end, MAX_LONGITUDE);
+        }
+    }
+
+    private static boolean edgeIntersectsTestEdge(Point arcStart, Point arcEnd, Point longitudeStart, Point longitudeEnd)
+    {
+        // y of longstart must be lower than y of longend
+        double longitude = longitudeStart.getX();
+        boolean arcCrossesLongitude = edgeCrossesLongitude(arcStart.getX(), arcEnd.getX(), longitude);
+
+        if (arcCrossesLongitude) {
+            double edgeLatitude = getArcLatitudeAtLongitude(arcStart, arcEnd, longitude);
+            boolean inRange = liesInRange(edgeLatitude, longitudeStart.getY() - 0.5, longitudeEnd.getY() + 0.5);
+            return inRange;
+        }
+        else {
+            return false;
+        }
+    }
+
+    private static double getArcLatitudeAtLongitude(Point start, Point end, double longitude)
+    {
+        SphericalExcessCalculator calculator = new SphericalExcessCalculator(start);
+        return calculator.computeExtrapolatedLatitudeThroughPoint(end, longitude);
+    }
+
+    private static Point getExternalPoint(double longitude)
+    {
+        return new Point(longitude, 90.0);
+    }
+
+    private static int getNumberIntersectionsWithPolygon(Polygon polygon, Point point)
+    {
+        int intersectionSum = 0;
+        Point testEdgeStart = point;
+        Point testEdgeEnd = getExternalPoint(point.getX());
+        if (point.getY() > testEdgeEnd.getY()) {
+            testEdgeStart = testEdgeEnd;
+            testEdgeEnd = point;
+        }
+
+        int numPaths = polygon.getPathCount();
+        for (int i = 0; i < numPaths; i++) {
+            int pathStartIndex = polygon.getPathStart(i);
+            int pathEndIndex = polygon.getPathEnd(i);
+            for (int j = pathStartIndex; j < pathEndIndex - 1; j++) {
+                Point edgeStart = polygon.getPoint(j);
+                Point edgeEnd = polygon.getPoint(j + 1);
+                if (edgeStart.getY() > edgeEnd.getY()) {
+                    Point temp = edgeStart;
+                    edgeStart = edgeEnd;
+                    edgeEnd = temp;
+                }
+                if (edgeStart.equals(point) || edgeEnd.equals(point)) {
+                    return 1;
+                }
+                if (edgeIntersectsTestEdge(edgeStart, edgeEnd, testEdgeStart, testEdgeEnd)) {
+                    intersectionSum++;
+                }
+            }
+            // check last edge that closes the loop back at the starting point
+            Point lastEdgeStart = polygon.getPoint(pathEndIndex - 1);
+            Point lastEdgeEnd = polygon.getPoint(pathStartIndex);
+            if (lastEdgeStart.equals(point) || lastEdgeEnd.equals(point)) {
```

Comment:
This check is not needed. All vertexes were already checked inside the for loop.